### ISSUE TITLE
Back button goes surprisingly far back in single file mode

### ISSUE THREAD

**Issue by danvk** [Reviewable/Reviewable#116]
When I do this sequence:

- Come to Reviewable for a large review from GitHub which puts you into single file mode
- Click on a few files to review the pull request
- Click the back button

I'd expect the back button to take you back through the files you've looked at. Instead, it takes you all the way back to GitHub.

**Comment by pkaminski** [comment]
Yeah, there's a couple reasons for that:
- It would be inconsistent to sometimes have a click in the file matrix / drop-down navigator affect the history and sometimes not.  And what about the various cycling actions (next file, next discussion, etc.) -- should those stack history too?
- At the moment, the auto-single-file mode doesn't change the URL when a new file is focused so stacking history would be confusing.  I'll probably combine the old-style and new-style single file modes at some point, though.

I guess I'm agreeing that this could be done better, but I punted on thinking it through in favor of getting the feature out sooner.  :blush:

**Comment by danvk** [comment]
There's definitely some subjectivity here. My intuition is that changing files in single page mode should be considered a history event, but opening/closing a comment or going to a new discussion in the same file should not. Going to a new file feels like you're navigating to a new page (especially with the file switcher hidden above the fold). The other actions don't.

**Comment by pkaminski** [comment]
What about changing diff bounds on a file?  And is it OK for history to behave differently in multi-file and single-file mode, especially considering that the mode can switch while you're looking at the page, if e.g. a new commit with more files comes in?  This will require more thought and experimentation...  For now, not touching history is at least predictable even if not always ideal.

**Comment by danvk** [comment]
Changing diff bounds could be a history event, though I don't feel strongly either way about it.

History events should only happen in response to user actions, not to incoming data. You wouldn't expect the back button to hide new emails from your inbox in Gmail, for example.

**Comment by pkaminski** [comment]
>History events should only happen in response to user actions, not to incoming data.

Yes, of course.  But my question is different: is it OK for the same user action to have a different effect on history depending on the mode the page is in?  Especially if this mode can change without user intervention?  And if the answer is "no, a given action should always have the same effect on history", then does that mean switching file in multi-file mode should also build history?

**Comment by danvk** [comment]
I'd be fine with navigating to a new file in all-files mode also being a history event. It feels like clicking an anchor to jump to another part of the page, which does go into your history by default.

**Comment by al-the-x** [comment]
Agreed on this point. There's lots of surprise with our team about how the back button works (or doesn't) in all modes. Navigating to a file, advancing to a comment, changing diffs should all be history events.

**Comment by seansfkelley** [comment]
I have a very fast oops-I-didn't-mean-to-click-that-hit-backspace reflex, so I hit this issue on pretty much every review. I'm ambivalent about how finely-tuned I want the history to be, but I would be happy if it just went back to the Reviewable home page for the review. That way I can re-correct my mistake quickly as I am still in Reviewable, but dropping me back into Github so I have to reload Reviewable really interrupts the flow (especially when I'm on a slow connection, which I am frequently).
